### PR TITLE
Add skip_absent to setcolorder

### PR DESCRIPTION
**Pull request by sluga**
Motivation: I sometimes want to ensure a certain order of columns but the set of columns is not fixed, i.e. a given column may or may not be present.

Now, working around that is trivial & so the benefit of this new argument isn't large, but it would be a small quality of life improvement & furthermore `setnames` already has the `skip_absent` argument, so it feels appropriate to add it to `setcolorder` as well.

### CODE REVIEW

**Review by MichaelChirico — COMMENTED** [review]
please revert whitespace changes first to make the diff clearer

**Comment by MichaelChirico** [comment]
Hi @sluga, happy to review but I still see a lot of spurious whitespace diff in R/data.table.R, PTAL

**Comment by sluga** [comment]
Missed the indent change. Might not be a bad idea to include a .Rproj file with the correct whitespace config in the repo; I assume at least some contributors use RStudio, and I don't think the file would be a problem for those who don't use RStudio.

**Comment by MichaelChirico** [comment]
> Missed the indent change. Might not be a bad idea to include a .Rproj file with the correct whitespace config in the repo; I assume at least some contributors use RStudio, and I don't think the file would be a problem for those who don't use RStudio.

Thanks for the suggestion. Those files are ignored:

https://github.com/Rdatatable/data.table/blob/585ec52e28e173c59ae1879d9dc3ade5f9477d95/.gitignore#L23-L24

As are several related to other IDEs like Emacs.

Better might be putting in the .dev folder, like our recommended .Rprofile:

https://github.com/Rdatatable/data.table/blob/master/.dev/.Rprofile

There users can get a copy when cloning the repo and just copy it to the package directory.

I am using VScode these days so not sure what an Rproj looks like anymore.

File: R/data.table.R [hunk 1]
```diff
@@ -2681,14 +2681,21 @@ setnames = function(x,old,new,skip_absent=FALSE) {
...
+setcolorder = function(x, neworder=key(x), before=NULL, after=NULL,skip_absent=FALSE)  # before/after #4358
 {
   if (is.character(neworder) && anyDuplicated(names(x)))
     stopf("x has some duplicated column name(s): %s. Please remove or rename the duplicate(s) and try again.", brackify(unique(names(x)[duplicated(names(x))])))
   if (!is.null(before) && !is.null(after))
     stopf("Provide either before= or after= but not both")
   if (length(before)>1L || length(after)>1L)
     stopf("before=/after= accept a single column name or number, not more than one")
+  if (!isTRUEorFALSE(skip_absent))
+    stopf("skip_absent should be TRUE or FALSE")
```
Comment (MichaelChirico) [inline R/data.table.R:2693]: This is a pretty common error, better for translation would be `stopf("%s should be TRUE or FALSE", "skip_absent")`, but that's out of scope for this PR.

Comment (MichaelChirico) [reply R/data.table.R:2693]: #6066

File: R/data.table.R [hunk 2]
```diff
@@ -2681,14 +2681,21 @@ setnames = function(x,old,new,skip_absent=FALSE) {
...
+  if (skip_absent && is.character(neworder)){
+    neworder = intersect(neworder, colnames(x))
+  } else if (skip_absent && is.numeric(neworder)){
+    neworder = intersect(neworder, seq_along(x))
+  }
   neworder = colnamesInt(x, neworder, check_dups=FALSE)  # dups are now checked inside Csetcolorder below
```
Comment (MichaelChirico) [inline R/data.table.R:2699]: More general would be an argument `skip_absent=` to `colnamesInt`, then the change can be even easier:

```r
neworder = colnamesInt(x, neworder, check_dups=FALSE, skip_absent=skip_absent)
```

But I think that's better for follow-up.

File: R/data.table.R (hunk 1 above)
Comment (MichaelChirico) [reply R/data.table.R:2693]: #6067

File: R/data.table.R (hunk 2 above)
Comment (MichaelChirico) [reply R/data.table.R:2699]: #6067

File: inst/tests/tests.Rraw [hunk 3]
```diff
@@ -1567,6 +1567,14 @@ test(498.03, setcolorder(DT, 1, after=3), data.table(b=2, c=3, a=1))
...
 test(498.05, setcolorder(DT, 1, before=1, after=1), error="Provide either before= or after= but not both")
 test(498.06, setcolorder(DT, 1, before=1:2), error="before=/after= accept a single column name or number, not more than one")
+# skip_absent
+test(498.07, setcolorder(DT, skip_absent = 'TRUE'), error = 'TRUE or FALSE')
+test(498.08, setcolorder(DT, skip_absent =  1), error = 'TRUE or FALSE')
+test(498.09, setcolorder(DT, skip_absent =  c(TRUE, FALSE)), error = 'TRUE or FALSE')
+test(498.10, setcolorder(DT, c('d', 'c', 'b', 'a')), error = 'non-existing column')
+test(498.11, setcolorder(DT, c('d', 'c', 'b', 'a'), skip_absent = TRUE), data.table(c=3, b=2, a=1))
+test(498.12, setcolorder(DT, 4:1), error = 'non-existing column')
+test(498.13, setcolorder(DT, 4:1, skip_absent = TRUE), data.table(a=1, b=2, c=3))
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:1577]:
```suggestion
test(498.07, setcolorder(DT, skip_absent='TRUE'), error='TRUE or FALSE')
test(498.08, setcolorder(DT, skip_absent= 1), error='TRUE or FALSE')
test(498.09, setcolorder(DT, skip_absent=c(TRUE, FALSE)), error='TRUE or FALSE')
test(498.10, setcolorder(DT, c('d', 'c', 'b', 'a')), error='non-existing column')
test(498.11, setcolorder(DT, c('d', 'c', 'b', 'a'), skip_absent=TRUE), data.table(c=3, b=2, a=1))
test(498.12, setcolorder(DT, 4:1), error='non-existing column')
test(498.13, setcolorder(DT, 4:1, skip_absent=TRUE), data.table(a=1, b=2, c=3))
```

File: man/setcolorder.Rd [hunk 4]
```diff
@@ -9,12 +9,13 @@
...
 
 \usage{
-setcolorder(x, neworder=key(x), before=NULL, after=NULL)
+setcolorder(x, neworder=key(x), before=NULL, after=NULL,skip_absent=FALSE)
 }
 \arguments{
   \item{x}{ A \code{data.table}. }
   \item{neworder}{ Character vector of the new column name ordering. May also be column numbers. If \code{length(neworder) < length(x)}, the specified columns are moved in order to the "front" of \code{x}. By default, \code{setcolorder} without a specified \code{neworder} moves the key columns in order to the "front" of \code{x}. }
   \item{before, after}{ If one of them (not both) was provided with a column name or number, \code{neworder} will be inserted before or after that column. }
+   \item{skip_absent}{ \code{FALSE} (the default) will throw an error if neworder includes columns not present in \code{x}, \code{TRUE} will quietly drop such columns.}
```
Comment (MichaelChirico) [inline man/setcolorder.Rd:18]:
```suggestion
   \item{skip_absent}{ Logical, default \code{FALSE}. If \code{TRUE}, no error is thrown if \code{neworder} includes columns not present in \code{x}, which are silently dropped. }
```

**Comment by MichaelChirico** [comment]
Thanks! Let's merge #6068 first since it's basically ready, then would you mind quickly refactoring this PR to use the new functionality?

**Comment by MichaelChirico** [comment]
OK, #6068 is merged, would you like to have a go at using the new argument to `colnamesInt()`?

File: README.md [hunk 5]
```diff
@@ -14,6 +14,12 @@
 
 `data.table` provides a high-performance version of [base R](https://www.r-project.org/about.html)'s `data.frame` with syntax and feature enhancements for ease of use, convenience and programming speed.
 
+---
+
+**NEW:** Take part in the [data.table 2023 community survey](https://tinyurl.com/datatable-survey/) and help shape the future of the project! The survey closes on **December 1st**.
```
Comment (MichaelChirico) [inline README.md:19]: not sure where this came from, please revert it

File: R/data.table.R [hunk 6]
```diff
@@ -2684,15 +2684,16 @@ setnames = function(x,old,new,skip_absent=FALSE) {
...
 {
   if (is.character(neworder) && anyDuplicated(names(x)))
     stopf("x has some duplicated column name(s): %s. Please remove or rename the duplicate(s) and try again.", brackify(unique(names(x)[duplicated(names(x))])))
   if (!is.null(before) && !is.null(after))
     stopf("Provide either before= or after= but not both")
   if (length(before)>1L || length(after)>1L)
     stopf("before=/after= accept a single column name or number, not more than one")
-  neworder = colnamesInt(x, neworder, check_dups=FALSE)  # dups are now checked inside Csetcolorder below
+  neworder = colnamesInt(x, neworder, check_dups=FALSE, skip_absent=skip_absent)  # dups are now checked inside Csetcolorder below
+  neworder = setdiff(neworder, 0) # tests 498.11, 498.13 fail w/o this
```
Comment (MichaelChirico) [inline R/data.table.R:2696]: `setdiff()` is not right here, consider:

```r
setdiff(c(1, 1), 0)
# [1] 1
```

You should add a test where `skip_absent=TRUE` and `neworder` has some duplicates as a further regression test.

File: R/data.table.R [hunk 7]
```diff
@@ -2684,15 +2684,16 @@ setnames = function(x,old,new,skip_absent=FALSE) {
...
 {
   if (is.character(neworder) && anyDuplicated(names(x)))
     stopf("x has some duplicated column name(s): %s. Please remove or rename the duplicate(s) and try again.", brackify(unique(names(x)[duplicated(names(x))])))
   if (!is.null(before) && !is.null(after))
     stopf("Provide either before= or after= but not both")
   if (length(before)>1L || length(after)>1L)
     stopf("before=/after= accept a single column name or number, not more than one")
-  neworder = colnamesInt(x, neworder, check_dups=FALSE)  # dups are now checked inside Csetcolorder below
+  neworder = colnamesInt(x, neworder, check_dups=FALSE, skip_absent=skip_absent)  # dups are now checked inside Csetcolorder below
+  neworder = neworder[!neworder %in% 0] # tests 498.11, 498.13 fail w/o this
```
Comment (MichaelChirico) [inline R/data.table.R:2696]: can neworder have NA?

if not, let's do neworder!=0

if so, let's do !is.na(neworder) & neworder!=0

Comment (sluga) [reply R/data.table.R]: I don't think it can be NA but @Nj221102 should confirm.

Comment (Nj221102) [reply R/data.table.R]: @sluga Yup, we can't use NA when `neworder` contains numbers (error), but you can use it when columns in `neworder` are represented by strings. This inconsistency was also mentioned in #6070, but we didn't get any use case for fixing it. However, now that we can fix it for this PR, I think we should. @MichaelChirico, what do you think?

```
dt = data.table(a=1, b=2, d=3)
> colnamesInt(dt, c("a", NA), skip_absent = TRUE)
[1] 1 0
> colnamesInt(dt, c(1, NA), skip_absent = TRUE)
Error in colnamesInt(dt, c(1, NA), skip_absent = TRUE) : 
  argument specifying columns received non-existing column(s): cols[2]=-2147483648

```

Comment (MichaelChirico) [reply R/data.table.R]: @Nj221102 I think the question here is whether `colnamesInt()` can return `NA`

Comment (Nj221102) [reply R/data.table.R]: > @Nj221102 I think the question here is whether `colnamesInt()` can return `NA`
sorry i misunderstood the question.

As far as I know, there is no way for `colnamesint` to return NA because when it doesn't find a column, it just returns 0 in its place. Additionally, NA is also converted to 0 when `cols` is character and throws an error for NA when `cols` is numeric.

Comment (MichaelChirico) [reply R/data.table.R]: Thanks for confirming!

File: inst/tests/tests.Rraw [hunk 8]
```diff
@@ -1585,6 +1585,15 @@ test(498.03, setcolorder(DT, 1, after=3), data.table(b=2, c=3, a=1))
...
 test(498.06, setcolorder(DT, 1, before=1:2), error="before=/after= accept a single column name or number, not more than one")
+# skip_absent
+test(498.07, setcolorder(DT, skip_absent='TRUE'), error='TRUE or FALSE')
+test(498.08, setcolorder(DT, skip_absent= 1), error='TRUE or FALSE')
+test(498.09, setcolorder(DT, skip_absent= c(TRUE, FALSE)), error='TRUE or FALSE')
+test(498.10, setcolorder(DT, c('d', 'c', 'b', 'a')), error='non-existing column')
+test(498.11, setcolorder(DT, c('d', 'c', 'b', 'a'), skip_absent=TRUE), data.table(c=3, b=2, a=1))
+test(498.12, setcolorder(DT, 4:1), error='non-existing column')
+test(498.13, setcolorder(DT, 4:1, skip_absent=TRUE), data.table(a=1, b=2, c=3))
+test(498.14, setcolorder(DT, c(1, 1, 2, 3), skip_absent=TRUE), error='!=')
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:1596]: Please add a test for `setcolorder(DT, c('a', 'b', 'd'), skip_absent=TRUE)`.

The NEWS item currently reads like 'c' will be _dropped_ from the output, which would be bad -- I would expect the above to be equivalent to `setcolorder(DT, c("a", "b"))`.

**Comment by MichaelChirico** [comment]
Hi @sluga I think this is near complete, just the last ask for a new test if you don't mind & we can merge. Thanks!

**Comment by sluga** [comment]
Sorry for the wait @MichaelChirico, added tests & revised NEWS & help slightly.

**Comment by MichaelChirico** [comment]
Thanks @sluga LGTM! I've invited you to become a project Member -- that means you can create branches directly on this repo for future contributions which will facilitate collaboration.

I added you as contributor to the DESCRIPTION using your name as written here:

https://github.com/rdatatable-community/The-Raft/blob/bfd1954dbbd1301ac272caadf353a5a9ce590a97/posts/2024-02-25-survey_2023-aljaz_sluga/index.qmd#L3

Please feel free to file a follow-up PR with any correction!

**Review by MichaelChirico — APPROVED** [review]
